### ISSUE TITLE
In-product site configuration

### ISSUE THREAD

**Issue by sqs** [sourcegraph/sourcegraph#965]
Site configuration is how site admins configure things for their Sourcegraph instance:

- repository mirroring (from GitHub, GitLab, and other code hosts)
- user authentication providers
- email (SMTP credentials)
- and other things

The full list of things is in [site.schema.json](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/schema/site.schema.json).

There are 3 problems with this that we want to solve, in order of importance:

- When running a Sourcegraph cluster, you can't edit the site config on the web. You need to edit a k8s configmap and then wait for a reload (or manually kill pods).
- Sourcegraph can't automatically migrate the site config content when fields change. (For example, the `auth.saml` -> `auth.provider` -> `auth.providers` change was painful.) This is true in both sourcegraph/server and cluster, because the config file is "owned" by the site admin, not by Sourcegraph.
- Tech debt: Propagating frontend values to other services (such as the list of gitservers) is weird.
- (Out of scope right now) Site config contains secrets, such as TLS keys, GitHub tokens, etc., that are sourced from central config management systems and should not be present in unencrypted form on disk (for some customers).

The solution must abide by the following:

- Currently a site admin on sourcegraph/server can make a change (on sourcegraph/server, and on cluster when we solve the 1st problem) that takes down their site, such as changing `appURL`. Right now, they can fix the problem by manually editing the site config file, but when we solve the 2nd problem, this might no longer be possible. This needs to not become a problem.
- It is OK to make certain fields in site configuration harder to edit (such as `appURL` and `auth.providers`) in exchange for making most others easier to edit.

Proposed solution:

1. Move site config that can't take down the site (i.e., all except `appURL`, `auth.providers`, etc.) out of the site config file and into "core site config", and make both be stored in the DB.
1. Make a new management console UI that allows editing the core config (the site admin page in-app will only allow editing the site config, the management console UI must be used to edit core config).
1. The management console UI has extra work needed around authentication, TBD at a later date once initial management console UI is complete.
1. (later?) Get https://github.com/sourcegraph/sourcegraph/pull/929 finished to remove more things from the "dangerous site config".

Decisions:

- "Deployment config" (`PG*`, `REDIS_*`, `SRC_GITSERVERS`) is renamed to "service connections".

Levels:

- User settings
- Org settings
- Global settings
- Site configuration
- Core configuration -> Critical configuration
- Deployment configuration -> Cluster service connections

Discussion:

- Is it important to store the "dangerous" site config outside the database? That lets you store "deployment configuration" in the config file as well.
  - Deployment configuration: `PG*`, `REDIS_*`, which gitserver instances to talk to

How other products solve this:

- GitLab's `gitlab.rb` config file 
  - GitLab [deciding on configurations](https://about.gitlab.com/handbook/product/#deciding-on-configurations): "When we have no choice, the secondary priority is to configure something in the GitLab interface. A configuration should only appear in a file (gitlab.rb or gitlab.yml) as a last resort."
- GitHub Enterprise's ini file

> [Draft announcement (from before kickoff)](https://github.com/sourcegraph/about/pull/36)

### Project status:

- Last week I moved significant portions of #966 into smaller PRs that I can land tomorrow with a strong guarantee of no regressions. (see PRs linked to from that one)
- I still need to address some TODOs in #966 before it itself can be merged, I am hoping to have that completed by Tue/Wed.
- The management-console itself still needs a UX (even if primitive / just a textarea for now). I am hoping to have the primitive version done late Wed / mid Thur.

**Comment by sqs** [comment]
FYI @slimsag I added this constraint on the solution:

> If someone is just running Sourcegraph on localhost using sourcegraph/server and does not need to change any critical config values from the defaults (which are the same as today, specifically defaulting to the builtin username-password auth provider), they should not need to access the management console UI at all. (This lets us retain the nice flow to try Sourcegraph of "run a single docker run command and go to http://localhost:7080").

and rewrote these 2 bullets in the proposed solution:

> - In the management console UI, add a Monaco editor with a JSON Schema for the critical config.
> - Figure out the authentication story for the management console UI. Use a simple password, make its port only accessible by using kubectl port-forward and similar for sourcegraph/server, etc.?

**Comment by slimsag** [comment]
Update:

- Most significant work is complete in https://github.com/sourcegraph/sourcegraph/pull/966 excluding the actual web UI for the management console.
- This week I will be trying to merge as much of that change as I can, by factoring some parts out into smaller-in-scope PRs (the PR has gotten too large to faithfully review/merge).
- Management console web UI expected to be out Wed/Thurs.

**Comment by slimsag** [comment]
@sqs do you have thoughts around this part:

> Figure out the authentication story for the management console UI. Use a simple password, make its port only accessible by using kubectl port-forward and similar for sourcegraph/server, etc.?

**Comment by slimsag** [comment]
Update:

- Last week I moved significant portions of #966 into smaller PRs that I can land tomorrow with a strong guarantee of no regressions. (see PRs linked to from that one)
- I still need to address some TODOs in #966 before it itself can be merged, I am hoping to have that completed by Tue/Wed.
- The management-console itself still needs a UX (even if primitive / just a textarea for now). I am hoping to have the primitive version done late Wed / mid Thur.

**Comment by slimsag** [comment]
Got it. For new instances we can require admins create it when they create the admin account, but for users migrating from an old version, what kind of experience do we need? Is running some manual SQL query to set the password OK?

**Comment by sqs** [comment]
For 3.0-preview, it is OK for the migration path to be rough. We will add a smooth migration (for many things, not just this) in 3.0.

Here is a sketch of the flow here that I think will work. My intent is to make it work for all deployment types, for both instances newly spun up at 3.0-preview and for older instances, and both at the time of initial setup and later. I also want the  initial setup flow to NOT need to first go to the mgmt console, which would add another step and probably decrease conversion.

- Each time the frontend server starts up, it checks for the presence of the mgmt console password. If none exists (which will occur for new instances and for instances upgraded from pre-3.0-preview), it generates a new strong password. It stores the bcrypted form of this password AND the plaintext password itself in the DB.
- All site admins who view Sourcegraph are prompted (on the site admin overview page) with a box like "Management console password: asdf1234. Copy to clipboard and **don't show again**" and with a link to the management console
- Once any site admin clicks **Don't show again**, the plaintext password is cleared from the DB (the bcrypted form remains, of course).
- Document a way for site admins to manually reset the password at a psql prompt. The value that they set doesn't need to come from a PostgreSQL query (eg you might want to bcrypt the value in Python or something on the CLI and have them paste that value into an SQL query, instead of requiring/assuming installation of bcrypt for PostgreSQL).

The only other things that might be necessary for 3.0 (not 3.0-preview) are:

- A way to change the mgmt console password from the mgmt console itself
- A way to set the mgmt console password in an env var or via k8s secrets instead of in the DB

**Comment by slimsag** [comment]
> 1. A way to change the mgmt console password from the mgmt console itself
> 2. A way to set the mgmt console password in an env var or via k8s secrets instead of in the DB

@sqs If you can make a strong argument for your second point here, I am happy to hear it. I've considered both just now, however, and what I think:

1. You really should only use the mgmt console passwords we generate securely for you. i.e., passwords that are not >= 128 chars in length will be insecure in nature due to no auth rate limiting. This means we should not do #2 above.
2. The mgmt console UI will let you regenerate the password, but it'll be automatic / you don't get to pick one.

**Comment by sqs** [comment]
@slimsag 	Sounds good to me. :+1:

**Comment by slimsag** [comment]
Filed https://github.com/sourcegraph/sourcegraph/issues/1438 for the UI to regenerate the password, but backlogged it given the pgsql command will do fine in the meantime.

**Comment by slimsag** [comment]
UPDATE: The management console is merged and working on Sourcegraph.com and dogfood as expected. Migration was automagic, etc.

Issues with config in general which MUST be solved for 3.0:

- https://github.com/sourcegraph/sourcegraph/issues?q=is%3Aissue+is%3Aopen+label%3Afeature%3Asite-config+-label%3Aidealistic-milestone+milestone%3A3.0

Issues with config in general which are ideally solved for 3.0, but can slip to later:

- https://github.com/sourcegraph/sourcegraph/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+label%3Afeature%3Asite-config+milestone%3A3.0+

I am closing this issue in favor of the above :)

**Comment by nicksnyder** [comment]
@slimsag instead of using "idealistic-milestone" can you use "release-blocker" for the ones that are required?

**Comment by slimsag** [comment]
Yes, I tagged the required ones as release-blocker, but I've left the idealistic milestone because I do want to communicate that it is ideal but perhaps not possible
